### PR TITLE
FOGL-2329 Add filter pipeline endpoint fails when script is added (uploaded) to filter before its added to a south / north pipeline

### PR DESCRIPTION
**Pull request by ashish-jabble**

### CODE REVIEW

File: python/foglamp/services/core/api/filters.py [hunk 1]
```diff
@@ -451,6 +451,8 @@ def _delete_keys_from_dict(dict_del: Dict, lst_keys: List[str], deleted_values:
             if parent is not None:
                 if dict_del['type'] == 'JSON':
                     i_val = json.loads(dict_del[k]) if isinstance(dict_del[k], str) else json.loads(json.dumps(dict_del[k]))
+                elif dict_del['type'] == 'script':
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/filters.py]: add a comment why this is needed

File: python/foglamp/services/core/api/filters.py [hunk 2]
```diff
@@ -451,6 +451,8 @@ def _delete_keys_from_dict(dict_del: Dict, lst_keys: List[str], deleted_values:
...
+                    i_val = ""
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/filters.py]: why its set to "", can it be empty dict?

File: python/foglamp/services/core/api/filters.py [hunk 3]
```diff
@@ -484,20 +486,39 @@ def _delete_keys_from_dict(dict_del: Dict, lst_keys: List[str], deleted_values:
...
+                await cf_mgr.update_configuration_item_bulk(filter_cat_in_pipeline, deleted_values)
+            # Special case to handle script type
+            if script_filter_config:
+                import binascii
+                hex_data = binascii.hexlify(script_filter_config['value'].encode('utf-8')).decode('utf-8')
+                # Remove cat from cache
+                if filter_cat_in_pipeline in cf_mgr._cacheManager.cache:
+                    cf_mgr._cacheManager.remove(filter_cat_in_pipeline)
+
+                await cf_mgr.set_category_item_value_entry(filter_cat_in_pipeline, item_name, hex_data, script_filter_config['file'])
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/filters.py]: will this set filter_cat_in_pipeline back to cache?

**Review by praveen-garg — CHANGES_REQUESTED** [review]
though it looks good to me, but some code level documentation is required.

File: python/foglamp/services/core/api/filters.py (hunk 2 above)
Comment (ashish-jabble) [reply python/foglamp/services/core/api/filters.py]: configuration value always expect in string, irrespective of the data type

**Comment by amarendra-dianomic** [comment]
@ashish-jabble I am getting below error when I executed curl commands as mentioned in FOGL-2329:
```
Jan 29 15:47:56 asinha-ThinkPad-E450 bench[16009]: INFO: Starting management api on port 0.
Jan 29 15:47:56 asinha-ThinkPad-E450 bench[16009]: INFO: Starting south service...
Jan 29 15:47:57 asinha-ThinkPad-E450 bench[16009]: INFO: Loading south plugin Benchmark.
Jan 29 15:47:57 asinha-ThinkPad-E450 bench[16009]: DEBUG: loadPlugin:96: name=Benchmark, type=south, default config={"plugin" : { "description" : "Simulated data generation for Benchmark tests", "type" : "string", "default" : "Benchmark", "readonly" : "true" }, "numAssets" : { "description" : "Number of unique assets to simulate", "type" : "string", "default" : "1", "order": "2", "displayName": "Number Of Assets" }, "asset" : { "description" : "Asset name prefix", "type" : "string", "default" : "Random", "order": "1", "displayName": "Asset Name" } }
Jan 29 15:47:57 asinha-ThinkPad-E450 bench[16009]: DEBUG: loadPlugin:113: Added entry in pluginHandleMap={0x1f95380, 0x1f94ac0}
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Benchmark plugin config: { "key" : "bench", "description" : "", "value" : {"numAssets" : { "description" : "Number of unique assets to simulate", "type" : "string", "value" : "1", "default" : "1" }, "asset" : { "description" : "Asset name prefix", "type" : "string", "value" : "Random", "default" : "Random" }, "plugin" : { "description" : "Simulated data generation for Benchmark tests", "type" : "string", "value" : "Benchmark", "default" : "Benchmark" }} }
Jan 29 15:47:58 asinha-ThinkPad-E450 FogLAMP[15856] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Registered service instance id=9042c613-b3a9-4a84-9327-1a3af5b8ef61: <bench, type=Southbound, protocol=http, address=localhost, service port=None, management port=43101, status=1>
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Registered service 9042c613-b3a9-4a84-9327-1a3af5b8ef61.
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Registered configuration category bench, registration id 0cfe1f65-6c73-4885-ba14-53f4b622d893.
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: DEBUG: Interest in bench registered
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Registered configuration category benchAdvanced, registration id 6ff64b46-f5be-4e04-b8f7-8d2d3c28c20d.
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: DEBUG: Interest in benchAdvanced registered
Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Connect to storage on localhost:37675
Jan 29 15:49:01 asinha-ThinkPad-E450 bench[16009]: WARNING: Filter 'python35', called without a Python 3.5 script. Check 'script' item in 'bench_p' configuration. Filter has been disabled.
```

**Comment by ashish-jabble** [comment]
> @ashish-jabble I am getting below error when I executed curl commands as mentioned in FOGL-2329:
> 
> ```
> Jan 29 15:47:56 asinha-ThinkPad-E450 bench[16009]: INFO: Starting management api on port 0.
> Jan 29 15:47:56 asinha-ThinkPad-E450 bench[16009]: INFO: Starting south service...
> Jan 29 15:47:57 asinha-ThinkPad-E450 bench[16009]: INFO: Loading south plugin Benchmark.
> Jan 29 15:47:57 asinha-ThinkPad-E450 bench[16009]: DEBUG: loadPlugin:96: name=Benchmark, type=south, default config={"plugin" : { "description" : "Simulated data generation for Benchmark tests", "type" : "string", "default" : "Benchmark", "readonly" : "true" }, "numAssets" : { "description" : "Number of unique assets to simulate", "type" : "string", "default" : "1", "order": "2", "displayName": "Number Of Assets" }, "asset" : { "description" : "Asset name prefix", "type" : "string", "default" : "Random", "order": "1", "displayName": "Asset Name" } }
> Jan 29 15:47:57 asinha-ThinkPad-E450 bench[16009]: DEBUG: loadPlugin:113: Added entry in pluginHandleMap={0x1f95380, 0x1f94ac0}
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Benchmark plugin config: { "key" : "bench", "description" : "", "value" : {"numAssets" : { "description" : "Number of unique assets to simulate", "type" : "string", "value" : "1", "default" : "1" }, "asset" : { "description" : "Asset name prefix", "type" : "string", "value" : "Random", "default" : "Random" }, "plugin" : { "description" : "Simulated data generation for Benchmark tests", "type" : "string", "value" : "Benchmark", "default" : "Benchmark" }} }
> Jan 29 15:47:58 asinha-ThinkPad-E450 FogLAMP[15856] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Registered service instance id=9042c613-b3a9-4a84-9327-1a3af5b8ef61: <bench, type=Southbound, protocol=http, address=localhost, service port=None, management port=43101, status=1>
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Registered service 9042c613-b3a9-4a84-9327-1a3af5b8ef61.
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Registered configuration category bench, registration id 0cfe1f65-6c73-4885-ba14-53f4b622d893.
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: DEBUG: Interest in bench registered
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Registered configuration category benchAdvanced, registration id 6ff64b46-f5be-4e04-b8f7-8d2d3c28c20d.
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: DEBUG: Interest in benchAdvanced registered
> Jan 29 15:47:58 asinha-ThinkPad-E450 bench[16009]: INFO: Connect to storage on localhost:37675
> Jan 29 15:49:01 asinha-ThinkPad-E450 bench[16009]: WARNING: Filter 'python35', called without a Python 3.5 script. Check 'script' item in 'bench_p' configuration. Filter has been disabled.
> ```

https://scaledb.atlassian.net/browse/FOGL-2325?focusedCommentId=40401&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-40401

**Review by amarendra-dianomic — DISMISSED** [review]
The code is working per specs of FOGL-2329 but I am not able to get the filter applied while applying curl commands as mentioned in the ticket. However, this might be handled vide another ticket.

**Comment by ashish-jabble** [comment]
@amarendra-dianomic With the changes error is no more reproducible with given steps mentioned in [JIRA](https://scaledb.atlassian.net/browse/FOGL-2329). If you see filter working issue please create a new JIRA for that

**Comment by praveen-garg** [comment]
while testing this, consider manual cleanup of uploaded script files; FOGL-2349

**Pull request closed without merging**